frame/collective/src/lib.rs

```diff
@@ -355,30 +648,57 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
 		approved: bool,
 		seats: MemberCount,
 		voting: Votes<T::AccountId, T::BlockNumber>,
-		proposal: T::Hash,
-	) {
+		proposal_hash: T::Hash,
+		proposal: Option<<T as Trait<I>>::Proposal>,
+	) -> Weight {
+		let db = T::DbWeight::get();
+		let mut weight: Weight = 0;
 		if approved {
-			Self::deposit_event(RawEvent::Approved(proposal));
+			Self::deposit_event(RawEvent::Approved(proposal_hash));
 
 			// execute motion, assuming it exists.
-			if let Some(p) = ProposalOf::<T, I>::take(&proposal) {
+			if let Some(p) = proposal {
+				let dispatch_weight = p.get_dispatch_info().weight;
 				let origin = RawOrigin::Members(voting.threshold, seats).into();
-				let ok = p.dispatch(origin).is_ok();
-				Self::deposit_event(RawEvent::Executed(proposal, ok));
+				let result = p.dispatch(origin);
+				Self::deposit_event(RawEvent::Executed(proposal_hash, result.is_ok()));
+				weight = weight.saturating_add(
+					get_result_weight(result).unwrap_or(dispatch_weight) // P1
+				);
 			}
 		} else {
 			// disapproved
-			ProposalOf::<T, I>::remove(&proposal);
-			Self::deposit_event(RawEvent::Disapproved(proposal));
+			Self::deposit_event(RawEvent::Disapproved(proposal_hash));
 		}
 
 		// remove vote
-		Voting::<T, I>::remove(&proposal);
-		Proposals::<T, I>::mutate(|proposals| proposals.retain(|h| h != &proposal));
+		ProposalOf::<T, I>::remove(&proposal_hash);
+		Voting::<T, I>::remove(&proposal_hash);
+		let mut proposal_len = T::MaxProposals::get() as Weight;
```

Comment:
What is up with this line?

Reply:
Forgot that mutate returns the value of the closure. :-D